src/python/pants/backend/experimental/python/lockfile.py

```diff
@@ -259,5 +291,50 @@ async def generate_all_tool_lockfiles(
     return ToolLockGoal(exit_code=0)
 
 
+BEGIN_LOCKFILE_HEADER = b"# --- BEGIN LOCKFILE METADATA: DO NOT EDIT OR REMOVE ---"
+END_LOCKFILE_HEADER = b"# --- END LOCKFILE METADATA ---"
+
+
+# Lockfile metadata for headers
+def lockfile_metadata(invalidation_digest: str) -> bytes:
+    """Produces a metadata bytes object for including at the top of a lockfile.
+
+    Currently, this only consists of an invalidation digest for the file, which is used when Pants
+    consumes the lockfile during builds.
+    """
+    return (
+        b"""
+%(BEGIN_LOCKFILE_HEADER)b
+# invalidation digest: %(invalidation_digest)s
+%(END_LOCKFILE_HEADER)b
+    """
+        % {
+            b"BEGIN_LOCKFILE_HEADER": BEGIN_LOCKFILE_HEADER,
+            b"invalidation_digest": invalidation_digest.encode("ascii"),
+            b"END_LOCKFILE_HEADER": END_LOCKFILE_HEADER,
+        }
+    ).strip()
+
+
+def read_lockfile_metadata(contents: bytes) -> dict[str, str]:
+    """Reads through `contents`, and returns the contents of the lockfile metadata block as a
+    dictionary."""
```

Comment:
Might want to wait to add this until it's in use. For example, this is going to need logic for what to do if the metadata is missing.

If you do still include it, a dedicated test would be great.

Reply:
(To be clear -- the content of `lockfile_metadata` shouldn't really matter as long as the data is retrievable from `read_lockfile_metadata`, so that's what I'm going to test)

Reply:
See https://github.com/pantsbuild/pants/pull/12427/files#diff-c88052d4bb845ecf33d278ef32987b89a2ee1de5fde6a9284f6ffb81bc912ee5R16-R20 for a test that does just that